libs/iblprefilter/src/IBLPrefilterContext.cpp

```diff
@@ -0,0 +1,387 @@
+
+    if (outReflectionsTexture == nullptr) {
+        outReflectionsTexture = createReflectionsTexture({});
+    }
+
+    const uint8_t inputLevels = environmentCubemap->getLevels();
+    const uint32_t sampleCount = mSampleCount;
+    const float linear = options.hdrLinear;
+    const float compress = options.hdrMax;
+    const uint8_t levels = outReflectionsTexture->getLevels();
+    uint32_t dim = outReflectionsTexture->getWidth();
+
+    TextureSampler environmentSampler;
+    environmentSampler.setMagFilter(SamplerMagFilter::LINEAR);
+    environmentSampler.setMinFilter(SamplerMinFilter::LINEAR_MIPMAP_LINEAR);
+
+    mi->setParameter("environment", environmentCubemap, environmentSampler);
+    mi->setParameter("compress", float2{ linear, compress });
+
+
+    if (options.generateMipmap) {
+        // We need mipmaps for prefiltering
+        environmentCubemap->generateMipmaps(engine);
+    }
+
+    RenderTarget::Builder builder;
+    builder.texture(RenderTarget::AttachmentPoint::COLOR0, outReflectionsTexture)
+           .texture(RenderTarget::AttachmentPoint::COLOR1, outReflectionsTexture)
+           .texture(RenderTarget::AttachmentPoint::COLOR2, outReflectionsTexture);
+
+
+    for (size_t lod = 0; lod < levels; lod++) {
+        SYSTRACE_NAME("computeFilterLOD");
+        const float omegaP = (4.0f * f::PI) / float(6 * dim * dim);
+        const float perceptualRoughness = lodToPerceptualRoughness(saturate(lod / (levels - 1.0f)));
+        const float roughness = perceptualRoughness * perceptualRoughness;
+
+        uint32_t effectiveSampleCount = sampleCount;
+        if (lod == 0) {
+            effectiveSampleCount = 1;
+        }
+
+        // compute the coefficients
+        half4* const p = (half4*)engine.streamAlloc(effectiveSampleCount * sizeof(half4));
+
+        float weight = 0.0f;
+        for (size_t i = 0; i < effectiveSampleCount; i++) {
+            const float2 u = hammersley(uint32_t(i), 1.0f / float(effectiveSampleCount));
+            const float3 H = hemisphereImportanceSampleDggx(u, roughness);
+            const float NoH = H.z;
+            const float NoH2 = H.z * H.z;
+            const float NoL = saturate(2 * NoH2 - 1);
+            const float3 L(2 * NoH * H.x, 2 * NoH * H.y, NoL);
+
+            const float pdf = DistributionGGX(NoH, roughness) / 4;
+            const float invOmegaS = effectiveSampleCount * pdf;
+            const float l = 1.0f - log4(invOmegaS) - log4(omegaP);
+            const float mipLevel = roughness == 0.0f ? 0.0f : clamp(l, 0.0f, inputLevels - 1.0f);
+
+            weight += NoL; // note: L.z == NoL
+            p[i] = { L, mipLevel };
+        }
+
+        // upload the coefficients
+        weights->setImage(engine, 0, {
+                p, effectiveSampleCount * sizeof(half4),
+                PixelDataFormat::RGBA, PixelDataType::HALF });
+
+        mi->setParameter("kernel", weights, TextureSampler{ SamplerMagFilter::NEAREST });
```

Comment:
I would maybe rename either `kernel` or `weights` to match the other.

Reply:
sounds good